src/main/java/de/uniba/dsg/serverless/provider/google/GoogleLogHandler.java

```diff
@@ -0,0 +1,185 @@
+		this.functionName = functionName;
+		this.startTime = startTime;
+		this.endTime = endTime;
+	}
+
+	private List<LogEntry> extractLogEntryFromPage(Page<LogEntry> entries) {
+		List<LogEntry> logEntries = new ArrayList<>();
+		for (LogEntry entry : entries.iterateAll()) {
+			logEntries.add(entry);
+		}
+		return logEntries;
+	}
+
+	private List<LogEntry> executeListLogEntries() throws SeMoDeException {
+
+		List<LogEntry> logEntries = new ArrayList<>();
+
+		// project is inferred from the environment
+		LoggingOptions options = LoggingOptions.getDefaultInstance();
+
+		try (Logging logging = options.getService()) {
+
+			String filter = "resource.type=\"cloud_function\" AND " + "resource.labels.function_name = \""
+					+ this.functionName + "\" AND " + "timestamp > \"" + this.startTime.format(QUERY_DATE_FORMATTER)
+					+ "\" AND " + "timestamp < \"" + this.endTime.format(QUERY_DATE_FORMATTER) + "\"";
+
+			EntryListOption entryFilterOption = EntryListOption.filter(filter);
+
+			Page<LogEntry> entries = logging.listLogEntries(entryFilterOption);
+			logEntries.addAll(this.extractLogEntryFromPage(entries));
+
+			// further pages with log entries
+			while (entries.hasNextPage()) {
+				entries = entries.getNextPage();
+				logEntries.addAll(this.extractLogEntryFromPage(entries));
+			}
+
+			return logEntries;
+		} catch (Exception e) {
+			throw new SeMoDeException("An error occured during accessing the stackdriver API", e);
+		}
+	}
+
+	/**
+	 * Different as on AWS and Azure. The key in the returned map is typically the
+	 * platform execution id. At google's platform, this key is not accessible and
+	 * therefore a self-assigned key within the cloud function in form of a uuid is
+	 * used to achieve the mapping from platform executions and local REST calls.
+	 */
+	@Override
+	public Map<String, WritableEvent> getPerformanceData() throws SeMoDeException {
+
+		List<LogEntry> logEntries = this.executeListLogEntries();
+
+		Map<String, List<LogEntry>> eventMap = this.generateEventMapByExecutionId(logEntries);
+
+		return this.generatePerformanceDataMap(eventMap);
+	}
+
+	private Map<String, WritableEvent> generatePerformanceDataMap(Map<String, List<LogEntry>> eventMap) {
+
+		Map<String, WritableEvent> performanceMap = new HashMap<>();
+		
+		for(String key : eventMap.keySet()) {
+			List<LogEntry> cohesiveEvent = eventMap.get(key);
+			PerformanceData data = this.extractPerformanceData(cohesiveEvent);
+			performanceMap.put(data.getRequestId(), data);
+		}
+		
+		return performanceMap;
+	}
+
+	// TODO: make the whole function more robust -> refactor
+	private PerformanceData extractPerformanceData(List<LogEntry> cohesiveEvent) {
+		
+		LocalDateTime startTime = LocalDateTime.MIN;
+		LocalDateTime endTime = LocalDateTime.MIN;
+		double preciseDuration = -1.0;
+		String platformId= "";
+		String instanceId = "";
+		
+		for(LogEntry entry : cohesiveEvent) {
+			Payload payload = entry.getPayload();
+			if(payload instanceof StringPayload) {
```

Comment:
Can be Payload be something else than a String? Should an exception be thrown if the payload is no string?

Reply:
The string can something else than a StringPayload like configuration message, but only the string payload messages are relevant for our purposes. There is no need to throw an exception here.